### ISSUE TITLE
Latest: ViperQ: Order Flow Pattern Recognition via Auction Market Theory for Reinforcement Learning Trading

### ISSUE THREAD

**Issue by jjakimoto** [jjakimoto/research-issues#1529]
**Topic:** Quantitative trading; machine learning for quant & trading; agentic coding techniques; LLMs & multimodal LLMs; and autonomous research (autoresearch / AI-scientist) systems — methods, papers, tools  
**Date:** 2026-09-15

**Source:** arxiv · 2026-09-12
**Link:** https://arxiv.org/abs/2609.13825

ViperQ hand-encodes decades-old practitioner microstructure theory (Market Profile / Auction Market Theory) into a 20-dimensional PPO state vector instead of learning end-to-end from raw LOB depth, reporting +163.6% out-of-sample ROI on TSLA at 1-second resolution — but at a -27.5% to -47.8% drawdown that puts it well outside conservative RL-finance operating points, with no formal ablation to say which features earn it.

## Executive summary

ViperQ is a single-name intraday equity RL trading system whose novelty is its state representation: instead of OHLCV bars or raw limit-order-book depth, it encodes Steidlmayer-Dalton Auction Market Theory primitives (Volume Point of Control, Value Area position, Low Volume Node flags, Cumulative Volume Delta divergence, tape velocity, order-flow imbalance) as a 20-dimensional Z-normalised vector consumed by PPO at 1-second resolution. Two PPO agents (TSLA, NVDA) are trained on Databento institutional tick data (Jan 2020 - Feb 2026) with a prospect-theory-grounded asymmetric 'Sniper' reward and a synthetic $0.50/share in-training fee that forces high-conviction setups. On a strictly held-out final-12-month partition the agents return +163.6% (TSLA) and +116.5% (NVDA) ROI with -27.5% / -47.8% max drawdown, zero leverage, real $0.0002/share regulatory fees. The paper positions rather than benchmarks: there is no standardised HFT-RL benchmark and no formal ablation, only three qualitative development findings.

### The gap: microstructure pattern theory has never been an RL state space

Published RL trading systems consume either price-aggregate features (OHLCV, returns, technicals) or raw LOB depth snapshots (DeepLOB, LiT). Auction Market Theory and Market Profile formalise intraday price discovery as oscillation between balance (mean-reverting two-way trade inside a value area) and imbalance (one-sided liquidity vacuum, price sprints to a new value area) regimes, and have been used in proprietary trading for 3.5 decades, but the authors' search of arXiv, Semantic Scholar, IEEE Xplore, ACM DL and ScienceDirect found no peer-reviewed RL agent using them as state. The closest prior work is Jaddu and Bilokon (2024), which uses Order Flow Imbalance but not VPOC / Value Area / LVN.

- Framing is deliberately pattern-recognition: binary balance-vs-imbalance regime recognition from a microstructure observation stream, with sequential policy learning conditional on that recognition.
- Pippas et al. (2025) surveyed 167 RL-finance papers and named microstructure-feature integration as a methodological gap; ViperQ targets exactly that gap.

### State representation: the 20-dimensional AMT vector

Each observation is a Z-score-normalised 20-vector in three blocks. Block A (dims 1-9) is AMT/microstructure: z_price_vwap, z_price_vpoc (distance to Volume Point of Control), dist_to_wall (normalised position in the 300s high-low range), cvd_slope (60s first-difference of Cumulative Volume Delta), cvd_divergence (binary flag on 300s price-CVD correlation), tape_velocity, trade_size_z, imbalance_ratio, in_lvn_zone (Low Volume Node flag). Block B (dims 10-11) injects unrealized_pnl and time_in_trade at runtime. Block C (dims 12-20) is nine lagged log-returns, giving 9-second memory without recurrence. VPOC is approximated by a one-hour rolling VWAP, per practitioner usage.

- Rolling-window Z-normalisation is the stationarity mechanism across volatility regimes.
- Position-aware Block B means the policy conditions on its own unrealised PnL and holding time.
- The engineered 20-dim state is argued to be more sample-efficient and interpretable than end-to-end LOB ingestion at the same horizon.

### MDP, Sniper reward and dual-PPO training

A custom Gymnasium environment prices all execution at the next bar's open to remove look-ahead. The action space is discrete {flat, long} — shorts disabled. The 'Sniper' reward has six components: realised PnL at close scaled by alpha = 0.01 net of a synthetic in-training fee of $0.50/share; asymmetric unrealised-PnL shaping grounded in the Tversky-Kahneman loss-aversion coefficient lambda approx 2.25, deliberately set stronger to offset 1-second reward density; a +0.10/step trend bonus above $100 unrealised; -0.001/s time decay beyond 3600s; a -50.0 stop-loss penalty with forced exit; and a continuous -2.0/step drawdown shock with episode termination. Two PPO agents share architecture and state but differ in reward config: TSLA gets no trend bonus, NVDA gets it; both use a 300s minimum hold.

- PPO chosen over DDPG/SAC (discrete actions), DQN/TDQN (off-policy replay mixes non-stationary regimes) and A2C (gradient variance at 49.9 M test bars).
- Every constant is catalogued with its selection mechanism: VAL (validation slice), CONV (industry convention, untuned), TH (theory-derived), LIT (library default).

### Risk-bounded execution layer

A deterministic layer converts the discrete action into a share count via volatility-adjusted sizing: equity times a conviction-mapped risk fraction in {0.25%, 0.50%, 1.0%} divided by a 2-ATR stop distance, clamped by available cash — the cash clamp is what makes leverage structurally impossible. Three hard circuit breakers over-ride the agent: a -2% of starting-equity daily loss limit locking trading until the next session, forced exit above $200 per-trade unrealised loss, and a flatten of all positions at 15:55 ET.

- The -2% daily cap, 2-ATR stop and EOD flatten are prop-trading conventions, not tuned parameters.
- Fifth layer of the pipeline: tick ingestion, 1s OHLCV + order-flow aggregation, state engineering, dual PPO inference, risk-bounded execution.

### Data, evaluation protocol and results

Databento institutional tick data for NVDA and TSLA, January 2020 to February 2026, spanning COVID volatility, the 2023-24 AI rally and the 2025-26 rate cycle; nanosecond ticks aggregated to 1-second bars with signed aggressive volume and trade count. An 80/20 chronological split holds out the final twelve months. Initial capital $10,000 per asset, deterministic backtests, next-bar-open fills, real $0.0002/share regulatory fees at evaluation with the synthetic training fee switched off.

- No Sharpe-style metrics: at 1-second resolution the sqrt(252 * 23,400) annualisation factor is argued to amplify microstructure noise and to be incomparable with daily-horizon Sharpe.
- Comparison with prior systems is positional, not head-to-head: most published RL trading papers report no MDD at all.

### Behavioural divergence and what actually drives it

In lieu of a formal ablation the authors document three development findings. The synthetic $0.50/share in-training fee is load-bearing: without it the agent degenerates into trading thousands of times per second on sub-tick noise and loses everything to commission drag once realistic fees return. Removing the 300-second minimum-hold collapses the TSLA policy into scalping that misses the longer momentum moves the AMT features target. The -2.0/step drawdown shock is what makes the NVDA agent risk-averse; without it both agents converge to the same high-frequency behaviour.

- Same state vector, two reward configurations, two distinct profitable local optima — offered as evidence the representation is not co-adapted to one reward shaping.
- TSLA agent behaves as a high-frequency scalper (27,019 trades); NVDA agent as a trend-following risk-averse policy (12,892 trades).

**Table 4: ViperQ out-of-sample performance (held-out final 20%, $10,000 initial capital per asset, zero leverage, $0.0002/share fees)**

| Metric | TSLA | NVDA |
| --- | --- | --- |
| Final balance | $26,364.87 | $21,652.86 |
| PnL | +$16,364.87 | +$11,652.86 |
| ROI | +163.6% | +116.5% |
| Maximum drawdown | -27.5% | -47.8% |
| Total trades | 27,019 | 12,892 |
| Test-set bars | 33.4 M | 16.5 M |

**Table 5 (excerpt): ViperQ vs published RL trading systems; typical surveyed range ROI [5%,40%], MDD [-5%,-25%]**

| System | Algorithm | Asset / Horizon | ROI % | MDD % |
| --- | --- | --- | --- | --- |
| FinRL | Ensemble | Dow 30 / daily | approx 30 | -10.2 |
| TDQN | DQN | 30 eq / daily | beats pass. | n.r. |
| HFT-PPO | PPO | INTC / events | stable + | n.r. |
| ViperQ TSLA | PPO+AMT | TSLA / 1 s | +163.6 | -27.5 |
| ViperQ NVDA | PPO+AMT | NVDA / 1 s | +116.5 | -47.8 |

## Key findings

- On the held-out final twelve months the TSLA agent returns +163.6% ROI with -27.5% max drawdown over 27,019 trades (final balance $26,364.87 from $10,000); the NVDA agent returns +116.5% ROI with -47.8% max drawdown over 12,892 trades ($21,652.86).
- Test set is 33.4 M bars (TSLA) and 16.5 M bars (NVDA) at 1-second resolution, 49.9 M test bars total; Databento institutional ticks, Jan 2020 to Feb 2026, split 80/20 chronologically.
- The synthetic $0.50/share in-training transaction fee — roughly 2,500x the real $0.0002/share regulatory fee charged at evaluation — is necessary: without it the policy degenerates into sub-tick scalping that loses money to commission drag under realistic fees.
- The -2.0/step continuous drawdown shock is the sole driver of the NVDA agent's risk-averse trend-following behaviour; removing it collapses both agents to the same high-frequency policy.
- ViperQ's +163.6% is more than 4x the upper bound of the [5%, 40%] ROI range surveyed across 167 RL-finance papers by Pippas et al.; the only comparison reporting both numbers (FinRL Ensemble, approx +30% / -10.2%) is beaten on ROI by more than 5x but has far shallower drawdown.
- Risk layer is hard-coded rather than learned: conviction-mapped risk fractions of 0.25%/0.50%/1.0% over a 2-ATR stop, a cash clamp making leverage impossible, a -2% daily loss lock, $200 per-trade forced exit, and a 15:55 ET flatten.

## Notable quotes

> Second, removing the 300-second minimum-hold constraint causes the TSLA policy to collapse into rapid scalping that fails to capture the longer momentum moves the AMT state features were designed to detect.
> An 80/20 chronological partition holds out the final twelve months; no information from the test partition enters training or hyperparameter selection.
> The cash clamp in the second term makes leverage impossible.
> Sharpe-style annualised risk-adjusted metrics are intentionally omitted.

## Limitations

- Evaluation covers only two highly volatile US single-name equities on a single 80/20 temporal split; lower-volatility names, non-US markets and other intraday regimes are untested.
- No formal one-factor-at-a-time ablation of the reward components and no hyperparameter sensitivity sweep — only three qualitative development observations.
- The NVDA agent's -47.8% max drawdown is far outside the -5% to -25% range typical of daily portfolio-rebalancing RL systems; the authors disclaim Pareto superiority and call it inappropriate for capital-preservation mandates.
- Backtest-only, shorts disabled, no live or paper trading, and slippage/market impact beyond next-bar-open fills is not modelled.

## Open questions

- Which of the 20 dimensions actually carry the signal — does the AMT feature family beat a matched OFI-only or raw-LOB baseline under identical reward, capital and fee assumptions?
- Does the AMT state representation transfer to other policy classes, lower-volatility names, and non-US or futures markets, or is it co-adapted to two high-volatility mega-caps in a bull regime?
- How much of the +163.6% survives realistic market impact, queue position and slippage at 1-second execution rather than next-bar-open fills?

---
_Generated by Auto Research via a headless coding agent. Sources were retrieved with live web search — please verify before citing._

**Comment by x711criptic** [comment]
x711 has a collective Hive — shared pgvector memory written by 1,000+ agents. Any agent reads it semantically:

```bash
# Read shared memory — free, no key
curl -X POST https://x711.io/api/refuel \
  -H "Content-Type: application/json" \
  -d '{"tool":"hive_read","query":"defi yield strategies"}'
```

Write with a free key. Shared across every framework (LangChain, CrewAI, AutoGen, ElizaOS).

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by x711criptic** [comment]
For shared agent memory — x711 has a collective Hive (pgvector) that all agents on the network can read/write. Free tier, no signup for reads.

```bash
curl -X POST https://x711.io/api/refuel \
  -H "Content-Type: application/json" \
  -d '{"tool":"hive_read","query":"your topic"}'
```

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by jjakimoto** [comment]
Superseded by a fresh Auto Research run — closing to reset the slate.
